src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java

```diff
@@ -1197,4 +1197,31 @@ public InputStream getFileContent(BitbucketSCMFile file) throws IOException, Int
         }
         return content;
     }
+
+    @Restricted(ProtectedExternally.class)
+    protected CloseableHttpResponse executeMethod(HttpRequestBase httpMethod)
+            throws IOException {
+        CloseableHttpClient client = getHttpClient(httpMethod);
+        CloseableHttpResponse response = client.execute(httpMethod, context);
+        while (response.getStatusLine().getStatusCode() == TOO_MANY_REQUESTS_HTTP_STATUS) {
+            response.close();
+            httpMethod.releaseConnection();
+
+            /*
+             * TODO: When bitbucket starts supporting rate limit expiration time, remove 5
+             * sec wait and put code to wait till expiration time is over. It should also
+             * fix the wait for ever loop.
+             */
+            LOGGER.fine("Bitbucket server API rate limit reached, sleeping for 5 sec then retry...");
+            try {
+                Thread.sleep(REQUEST_WAIT_TIME_DEFAULT);
+            } catch (InterruptedException e) {
```

Comment:
Why do we catch this exception?

Reply:
> Why do we catch this exception?

It is an unchecked exception, should we through it and let the caller handle it?

Reply:
I would say so. Or alternatively we can at least break out of the while loop

Reply:
It is an unchecked exception, should we through it and let the caller handle it?

Reply:
Handled by throwing exception to the caller as caller already handles it.